src/uwtools/rocoto.py

```diff
@@ -25,20 +35,118 @@ def _add_jobname(tree: dict) -> None:
+    return values
+
+
+def _rocoto_schema_xml() -> DefinitePath:
+    """
+    The path to the file containing the schema to validate the XML file against.
+    """
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        return path / "schema_with_metatasks.rng"
+
+
+def _rocoto_schema_yaml() -> DefinitePath:
+    """
+    The path to the file containing the schema to validate the YAML file against.
+    """
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        return path / "rocoto.jsonschema"
+
+
+def _rocoto_template_xml() -> DefinitePath:
+    """
+    The path to the file containing the Rocoto workflow document template to render.
+    """
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        return path / "rocoto.jinja2"
+
+
+def _write_rocoto_xml(
+    config_file: OptionalPath,
+    rendered_output: DefinitePath,
+) -> None:
+    """
+    Render the Rocoto workflow defined in the given YAML to XML.
+
+    :param config_file: Path to YAML input file.
+    :param rendered_output: Path to write rendered XML file.
+    """
+
+    values = _add_jobname_to_tasks(config_file)
 
     # Render the template.
-    template = J2Template(values=values.data, template_path=input_template)
-    template.dump(output_path=rendered_output)
+    template = J2Template(values=values.data, template_path=_rocoto_template_xml())
+    template.dump(output_path=str(rendered_output))
+
+
+# Public functions
+def realize_rocoto_xml(
+    config_file: OptionalPath,
+    rendered_output: DefinitePath,
+) -> bool:
+    """
+    Realize the Rocoto workflow defined in the given YAML as XML. Validate both the YAML input and
+    XML output.
+
+    :param config_file: Path to YAML input file.
+    :param rendered_output: Path to write rendered XML file.
+    """
+
+    # Validate the YAML.
+    if uwtools.config.validator.validate_yaml(
+        config_file=config_file, schema_file=_rocoto_schema_yaml()
+    ):
+        _add_jobname_to_tasks(config_file)
+        # Render the template to a temporary file.
+        with tempfile.NamedTemporaryFile(delete=False) as temp_file:
+            _write_rocoto_xml(
+                config_file=config_file,
+                rendered_output=temp_file.name,
+            )
+            # Validate the XML.
+            if validate_rocoto_xml(input_xml=temp_file.name):
+                # If no issues were detected, save temp file and report success.
+                shutil.move(temp_file.name, rendered_output)
+                return True
+        logging.error("Rocoto validation errors identified in %s", temp_file.name)
+        return False
+    logging.error("YAML validation errors identified in %s", config_file)
+    return False
+
+
+def validate_rocoto_xml(input_xml: OptionalPath) -> bool:
+    """
+    Given a rendered XML file, validate it against the Rocoto schema.
+
+    :param input_xml: Path to rendered XML file.
```

Comment:
Would be nice to have a `:return:` docstring clause describing the meaning of the return value.